crates/stackable-webhook/src/lib.rs

```diff
@@ -200,19 +168,59 @@ impl WebhookServer {
         };
 
         // select requires Future + Unpin
-        pin_mut!(future_server);
-        pin_mut!(future_signal);
-
-        futures_util::future::select(future_server, future_signal).await;
+        tokio::pin!(future_server);
+        tokio::pin!(future_signal);
+
+        tokio::select! {
+            res = &mut future_server => {
+                // If the server future errors, propagate the error
+                res?;
+            }
+            _ = &mut future_signal => {
+                tracing::info!("shutdown signal received, stopping server");
+            }
+        }
 
         Ok(())
     }
 
-    /// Runs the webhook server by creating a TCP listener and binding it to
-    /// the specified socket address.
     async fn run_server(self) -> Result<()> {
         tracing::debug!("run webhook server");
 
-        self.tls_server.run().await.context(RunTlsServerSnafu)
+        let Self {
+            options,
+            mut webhooks,
+            tls_server,
+            mut cert_rx,
+            // initial_reconcile_tx,
+        } = self;
+        let tls_server = tls_server
+            .run()
+            .map_err(|err| WebhookServerError::RunTlsServer { source: err });
+
+        let cert_update_loop = async {
+            loop {
+                while let Some(cert) = cert_rx.recv().await {
+                    // The caBundle needs to be provided as a base64-encoded PEM envelope.
+                    let ca_bundle = cert
+                        .to_pem(LineEnding::LF)
+                        .context(EncodeCertificateAuthorityAsPemSnafu)?;
+                    let ca_bundle = ByteString(ca_bundle.as_bytes().to_vec());
+
+                    for webhook in webhooks.iter_mut() {
+                        webhook
+                            .handle_certificate_rotation(&cert, &ca_bundle, &options)
+                            .await
+                            .context(UpdateCertificateSnafu)?;
+                    }
+                }
+            }
+
+            // We need to hint the return type to the compiler
+            #[allow(unreachable_code)]
+            Ok(())
```

Comment:
We might be able to avoid this through the `!` (Never) type. But I guess that won't work with returning errors via `?`.

I am curious though.